src/ir/auth_logic/ast_construction_says_assertion_test.cc

```diff
@@ -61,47 +61,52 @@ INSTANTIATE_TEST_SUITE_P(
         SaysAssertionTestData({R"("UserA" says { currTimeIs(2022).
                                mayA(NotifyA) :- currTimeIs > 2021. })",
                                R"("UserA")",
-                               {"currTimeIs(2022)",
-                                "mayA(NotifyA) :- currTimeIs > 2021"}}),
+                               {"Fact(BaseFact(currTimeIs(2022))).",
```

Comment:
It looks like the existing `ToString` was meant to be human readable. With this PR, we are making it expose the AST hierarchy. Is that desirable?

IIRC, one (or the only?) reason we exposed the AST nodes like `Fact` and `BaseFact` was to test the visitor. With https://github.com/google-research/raksha/commit/9706fa433bf0281ee4f2e4829bb19c9139f1f911, we do not need `ToString` for testing the visitor, right? Did we decide that it is not possible to fix the ast visitor tests without `ToString`?

Reply:
This `ToString` function that I did not write and that was already in our repository and used for several tests that I also did not write is used for debugging and testing, and for both of these use-cases distinguishing between the AST nodes seems useful. For both of these use cases exposing the AST is important.

The souffle emitter is a way of converting the AST into strings that are souffle code. That emitter is distinct from this ToString and has a different purpose. It does not expose the AST.

As I tried to explain in the comment at the top of this PR, this `ToString` function defined in `ast_string.h` is distinct from the `ToString` methods inside `ast.h` [like this one](https://github.com/google-research/raksha/blob/main/src/ir/auth_logic/ast.h#L53). The whole point of this PR is to make the `ToString` function in `ast_string.h` sufficient for the uses of the method in `ast.h` so I can delete the methods in `ast.h` and use only `ast_string.h` later to remove this redundancy.

https://github.com/google-research/raksha/commit/9706fa433bf0281ee4f2e4829bb19c9139f1f911 did not remove the need to have a ToString method for the visitor test. You and I sitting together for a full hour were not able to think of a way to test the visitor with equality because in C++ there is no way to declare a vector of a class with a pure virtual method. So we still need it for that test. Also as mentioned in this comment, there were already tests in the code (that I did not write) that used ToString for testing.

Reply:
I understood that you were consolidating the `ToString` methods. I was actually wondering if the `ToString` in `ast_string.h` is meant to be human readable and if this change is removing this functionality. 

BTW, the use of `std::variant` in this PR gave me an idea to remove the dependency on `ToString` for ast visitor tests. (See  https://github.com/google-research/raksha/pull/734/files and the internal CL.)

Reply:
While I was at it, I went ahead and removed the `ToString` methods from Ast as well. (See https://github.com/google-research/raksha/pull/736 and the internal CL)

Reply:
> BTW, the use of std::variant in this PR gave me an idea to remove the dependency on ToString for ast visitor tests. (See https://github.com/google-research/raksha/pull/734/files and the internal CL.)

Ah nice that we have a way to fix it. I can't see the change though. When I look at that link it has 0 commits and 0 files changed... hmm...

Reply:
I thought that outputting the details of the AST was a good thing because this `ToSTring` is meant to help with debugging. If you strongly feel it should not output those details, I can change it the other way so that we can make forward progress though!

Reply:
I guess this print is also used in tests, so changing this print means the tests also have to change... OK I can make this toString output fewer details of the AST then

Reply:
Ah weirdly I have an e-mail with a reply from you in this thread, but it isn't showing up in this thread on the github webpage!